ticketbuyer: Add get/set config

tuxcanfly opened the pull request:
Depends on #228 

Added methods to get/set individual parameters of ticketbuyer config. Updated `AutoBuyerViewModel` from #228 to include methods corresponding to RPCs.

Refs #180

tuxcanfly committed:
ticketbuyer: Add AutoBuyerViewModel

tuxcanfly committed:
Add ViewModel methods for ticketbuyer RPCs

tuxcanfly committed:
Fix Account name -> int conversion

jrick commented on Paymetheus.Decred/Wallet/AutoBuyerProperties.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        public byte[] Passphrase { get; set; }
+        public uint Account { get; set; }
+        public long BalanceToMaintain { get; set; }
```
use Amount

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex);
```
please never do this. the caller must handle the exception, we can't do that here.

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex);
```
see above

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex);
```
see above

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            }
+        }
+
+        public async Task SetAccount(uint account)
```
use the Account type for accounts

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            try
+            {
+                var client = new TicketBuyerService.TicketBuyerServiceClient(_channel);
+                var req = new StartAutoBuyerRequest();
```
```
var req = new StartAutoBuyerRequest
{
    Passphrase = ByteString.CopyFrom(properties.Passphrase),
    // ...
};

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            {
+                var client = new TicketBuyerService.TicketBuyerServiceClient(_channel);
+                var req = new SetBalanceToMaintainRequest();
+                req.BalanceToMaintain = balance;
```
see syntax above

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex);
```
nope

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex);
```
nope

i'll stop commenting about these. go fix them all

tuxcanfly replied:
Ok, will do.

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            {
+                var client = new TicketBuyerService.TicketBuyerServiceClient(_channel);
+                var req = new SetMaxFeeRequest();
+                req.MaxFee = maxFee;
```
use the syntax above, do this for all requests please.

tuxcanfly replied:
Thanks, will do that.

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            }
+        }
+
+        public async Task SetMaxFee(long maxFee)
```
Amount

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            }
+        }
+
+        public async Task SetMaxPriceAbsolute(long maxPriceAbsolute)
```
Amount

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            }
+        }
+
+
```
extra newline

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            {
+                Console.WriteLine(ex);
+            }
+
```
extra newline

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            {
+                Console.WriteLine(ex);
+            }
+
```
extra newline

jrick commented on Paymetheus.Rpc/WalletClient.cs:
```diff
@@ -643,5 +644,201 @@ public void CancelRequests()
+            {
+                Console.WriteLine(ex);
+            }
+
```
extra newline

jrick commented on Paymetheus/ViewModels/AutoBuyerViewModel.cs:
```diff
@@ -0,0 +1,95 @@
+        public ICommand StartAutoBuyerCommand { get; }
+        public DelegateCommand StopAutoBuyerCommand { get; private set; }
+
+        private async void StartAutoBuyerAction()
```
you don't want to use async void here, use async Task, and change DelegateCommand to DelegateCommandAsync

jrick commented on Paymetheus/ViewModels/AutoBuyerViewModel.cs:
```diff
@@ -0,0 +1,95 @@
+        private async Task<AutoBuyerProperties> GetAutoBuyerProperties()
+        {
+            AutoBuyerProperties autobuyerProperites = await App.Current.Synchronizer.WalletRpcClient.AutoBuyerPropertiesAsync();
+            return autobuyerProperites;
```
simpler to write

`return await App.Current.Synchronizer.WalletRpcClient.AutoBuyerPropertiesAsync();`

or just remove this helper since it's just one line anyways

jrick commented on Paymetheus/ViewModels/AutoBuyerViewModel.cs:
```diff
@@ -0,0 +1,95 @@
+            return autobuyerProperites;
+        }
+
+        private async void SetAccoun(uint account)
```
misspelled and needs to use Account type

jrick replied:
Also there's no need to write setter functions, that's what properties are for.  Only properties can be bound to by the view, not functions.

tuxcanfly committed:
Address review items:

* Use Amount for amt fields - BalanceToMaintain, MaxPriceAbsolute, MaxFeePerKb
* Removve try, except in WalletClient methods

tuxcanfly committed:
Address review items

* Start/StopAutoBuyerAction- Use DelegateCommandAsync
* Simplify GetAutoBuyerProperties
* Remove AutoBuyerProperties Setters
* WalletClient - Simplify Request constructor
* Misc style fixes

tuxcanfly committed:
Address review items:

* AutoBuyerViewModel - Simplify Action methods

tuxcanfly commented:
Just noting here that `api.proto`  and methods/params need to be updated once the API in https://github.com/decred/dcrwallet/pull/598 is finalized.

tuxcanfly committed:
Add autobuyer checkbox, bind to write ini key

tuxcanfly committed:
Remember Auto Buyer checkbox state on startup

tuxcanfly committed:
Prompt  passphrase on startup is auto buyer is enabled

tuxcanfly committed:
WIP: Toggle autobuyer Start/Stop based on checkbox

tuxcanfly committed:
Add PassphraseDialog on initial AutoBuyer Start

tuxcanfly committed:
Added ticketbuyer fields

tuxcanfly committed:
Add ticketbuyer section to defaults.ini

tuxcanfly committed:
Save autobuyer config to defaults.ini

tuxcanfly committed:
Use AutoBuyerProperties for consistency

tuxcanfly committed:
Add Account, Validation

tuxcanfly committed:
Improve AutoBuyer Validation

tuxcanfly committed:
Add MaxPriceRelative

tuxcanfly committed:
Updated api.proto

tuxcanfly commented:
Updated `api.proto` to sync with dcrwallet. Working on passphrase validation for the startup dialog when autobuyer is enabled. Also working on updating ticketbuyer config when fields are edited.

Early testing, feedback and review welcome.

tuxcanfly committed:
Refactor PurchaseTicketsViewModel

tuxcanfly committed:
AutoBuyer - handle startup passphrase exception

tuxcanfly committed:
AutoBuyer - update config fields

tuxcanfly commented:
Added support for configuring individual ticketbuyer fields.

jrick commented:
Made a few comments in chat but adding here for tracking:

* I am seeing it call PurchaseTicketsViewModel.StartAutoBuyer from SynchronizerViewModel.OnSyncedWallet my first time running the code, and this is causing a NPE.  I have not enable the auto buyer at all and it should not be running this code.

* There was an "Auto buyer" selection added to the stakepool seletion dropdown menu.  This isn't correct, we want to allow the auto buyer to be used with any voting wallet, including stakepools.

* I find it very confusing that there is both an entry box for number of tickets to buy, as well as the auto buyer config settings, in the same list of settings.  These must be mutually exclusive.

tuxcanfly committed:
AutoBuyer - restore enabled check on startup

tuxcanfly committed:
AutoBuyer - separate config from stakepool settings

tuxcanfly committed:
AutoBuyer - Hide conflicting manual options

Also fix

* Don't trigger config updates when autobuyer is off
* Stop autobuyer when Enable AutoBuyer checkboxi is unchecked

tuxcanfly commented:
OK, addressed the above issues:

* Missed a check for whether ticketbuyer was enabled. Fixed.

* Removed menu entry and added a checkbox. This checkbox displays configuration options and a checkbox to toggle the ticketbuyer. This checkbox would probably be redesigned as a toggle button and placed more prominently (maybe beside the Purchase button) once the design is finalized, but I've kept it as normal for now.

* Updated to hide the tickets to purchase manually, when ticketbuyer checkbox is checked. It is restored when switching back to manual purchases.

tuxcanfly committed:
AutoBuyer - disable if passphrase was incorrect

tuxcanfly committed:
AutoBuyer - Handle failed start due to invalid passphrase

Also fix:

* Rename Enable AutoBuyer -> Show AutoBuyer options
* Separate ticketbuyer enabled vs config

tuxcanfly commented:
When a user enters a wrong passphrase on startup, we show an error message and disable the ticketbuyer so either they can re-enable it or correct the passphrase on next startup.

tuxcanfly committed:
AutoBuyer - update passphrase dialog text

Also update invalid passphrase err text

tuxcanfly committed:
AutoBuyer: Update error message

tuxcanfly commented:
@jrick  Need a way to trigger a callback if  use clicks the "cancel" button in `PassphraseDialogViewModel` Dialog, since we want to avoid setting ticketbuyer state to enabled. Is there a way to do that currently? I am guessing we need to add another Delegate command similar to ExecuteAction and pass it in.

jrick commented:
Correct, you have to bind it to a command

tuxcanfly committed:
AutoBuyer - Handle PassphraseDialog Cancel

tuxcanfly commented:
Thanks @jrick Added a cancel handler, and using it to prevent ticketbuyer state from being set to enabled on clicking "Cancel"

tuxcanfly committed:
AutoBuyer - Show notice informing that ticketbuyer only runs as long as wallet is running.

jrick reviewed (changes requested):
Alright got some more feedback.

When a stakepool is selected, the options for the pool fee address and the fee % need to be hidden.  These values must come from the stakepool.

When any of the auto buyer options are turned on, the ticket and split fees should be hidden.  Expiry too i think?

It would be useful if the rest of the auto buyer settings had tooltips with an explanation.  As is, someone new to the software won't have any clue what many of these options do.

The "Toggle Autobuyer" checkbox is also extremely confusing to me.  I want the purchase tickets button at the bottom to switch to an "enable autobuyer" button if that is what i am configuring.  It's also bad UX to display a dialog window if a checkbox is toggled.  When the autobuyer is already on, there needs to be a different button to turn it off so that tickets can be manually purchased again.

tuxcanfly committed:
AutoBuyer- toggle pool fields based on stake pool selection

tuxcanfly committed:
AutoBuyer - hide ticket, split fee and expiry on auto buyer

jrick commented:
Closing this PR.  May come back to this feature at a later time.
